model/map/src/main/java/org/keycloak/models/map/events/MapEventStoreProviderFactory.java

```diff
@@ -0,0 +1,107 @@
+import org.keycloak.models.map.common.AbstractMapProviderFactory;
+import org.keycloak.models.map.storage.MapStorage;
+import org.keycloak.models.map.storage.MapStorageProvider;
+import org.keycloak.models.map.storage.MapStorageProviderFactory;
+import org.keycloak.models.map.storage.MapStorageSpi;
+import org.keycloak.provider.EnvironmentDependentProviderFactory;
+import org.keycloak.provider.InvalidationHandler;
+
+import static org.keycloak.models.map.common.AbstractMapProviderFactory.MapProviderObjectType.REALM_BEFORE_REMOVE;
+import static org.keycloak.models.map.common.AbstractMapProviderFactory.uniqueCounter;
+import static org.keycloak.models.utils.KeycloakModelUtils.getComponentFactory;
+
+public class MapEventStoreProviderFactory implements AmphibianProviderFactory<EventStoreProvider>, EnvironmentDependentProviderFactory, EventStoreProviderFactory, InvalidationHandler {
+
+    public static final String PROVIDER_ID = AbstractMapProviderFactory.PROVIDER_ID;
+    private Config.Scope storageConfigScopeAdminEvents;
+    private Config.Scope storageConfigScopeLoginEvents;
+    private final String uniqueKey = getClass().getName() + uniqueCounter.incrementAndGet();
+
+
+    @Override
+    public void init(Config.Scope config) {
+        storageConfigScopeAdminEvents = config.scope(AbstractMapProviderFactory.CONFIG_STORAGE + "-admin-events");
+        storageConfigScopeLoginEvents = config.scope(AbstractMapProviderFactory.CONFIG_STORAGE + "-auth-events");
+    }
+
+    @Override
+    public void postInit(KeycloakSessionFactory factory) {
+    }
+
+    @Override
+    public EventStoreProvider create(KeycloakSession session) {
+        MapEventStoreProvider provider = session.getAttribute(uniqueKey, MapEventStoreProvider.class);
+        if (provider != null) return provider;
+
+        MapStorageProviderFactory storageProviderFactoryAe = (MapStorageProviderFactory) getComponentFactory(session.getKeycloakSessionFactory(),
+                MapStorageProvider.class, storageConfigScopeAdminEvents, MapStorageSpi.NAME);
+        final MapStorageProvider factoryAe = storageProviderFactoryAe.create(session);
+        MapStorage adminEventsStore = factoryAe.getStorage(AdminEvent.class);
+
+        MapStorageProviderFactory storageProviderFactoryLe = (MapStorageProviderFactory) getComponentFactory(session.getKeycloakSessionFactory(),
+                MapStorageProvider.class, storageConfigScopeLoginEvents, MapStorageSpi.NAME);
+        final MapStorageProvider factoryLe = storageProviderFactoryLe.create(session);
+        MapStorage loginEventsStore = factoryLe.getStorage(Event.class);
+
+        provider = new MapEventStoreProvider(session, loginEventsStore, adminEventsStore);
+        session.setAttribute(uniqueKey, provider);
+        return provider;
+    }
+
+    @Override
+    public void invalidate(KeycloakSession session, InvalidationHandler.InvalidableObjectType type, Object... params) {
+        if (type == REALM_BEFORE_REMOVE) {
```

Comment:
Should we do this? The legacy store doesn't do this, but I think we should.

Reply:
Related question: Should we do similar on Client/User and maybe other removals?

Reply:
let's keep it here, and only for realm